realign_bam_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/output/all_tissues/genome_bam
gtex_ids=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/data/other_references/editing_test_samples.txt
reference_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/data/edsite_references
reference_fasta=Homo_sapiens_assembly38_noALT_noHLA_noDecoy.fasta
dbsnp=All_20180418.vcf.gz
indels_mills=Mills_and_1000G_gold_standard.indels.hg38.sites.vcf
indels_decoy=Homo_sapiens_assembly38_1000genomes_decoy.known_indels.vcf
gene_intervals_bed=gencode.v47.merged.bed
exac_reference=small_exac_common_3.hg38.combined.vcf.gz
output_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/output/editing_test
code_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/scripts/processing
vcf_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/data/processed/vcfs
full_vcf_file=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/data/edsite_references/GTEx_Analysis_2021-02-11_v9_WholeGenomeSeq_953Indiv.SHAPEIT2_phased.vcf.gz
regenerate_all=true
step_size=1
max_array_size=1000
submit_on=sherlock