{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### lncRNAs quantified? relative to background\n",
    "1. construct backgorund gtf\n",
    "2. count reads with rnaseqc\n",
    "3. compare to see if lncRNAs are quantified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pandas'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mModuleNotFoundError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpandas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpd\u001b[39;00m\n\u001b[32m      2\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mseaborn\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01msns\u001b[39;00m\n\u001b[32m      3\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpyranges\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpr\u001b[39;00m\n",
      "\u001b[31mModuleNotFoundError\u001b[39m: No module named 'pandas'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import pyranges as pr\n",
    "import numpy as np\n",
    "import bisect\n",
    "\n",
    "from tqdm import tqdm\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### create background gtf\n",
    "\n",
    "bed files for exons and background are created with `scripts/analysis/create_matched_nongenic_null.sh`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = pd.read_csv(\"/oak/stanford/groups/smontgom/klawren/nph_gtex/data/tmp/macaques/exons_to_match.bed\", sep='\\t', header=None, names=['chr', 'start', 'end', 'exon_id', 'gene_id'])\n",
    "exons['length'] = exons['end'] - exons['start']\n",
    "exons = exons[~exons['chr'].isin(['chrM'])]\n",
    "\n",
    "# if introns are allowed\n",
    "#background = pd.read_csv(\"/home/klawren/oak/gtex/data/other_references/nongenic_null/background_regions.bed\", sep='\\t', header=None, names=['chr', 'start', 'end'])\n",
    "# if introns aren't allowed \n",
    "background = pd.read_csv(\"/oak/stanford/groups/smontgom/klawren/nph_gtex/data/tmp/macaques/intergenic_regions.bed\", sep='\\t', header=None, names=['chr', 'start', 'end'])\n",
    "\n",
    "background = background[background['chr'].isin(exons['chr'].unique())]\n",
    "background = background.sort_values(['chr', 'start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in gencode 47 gtf\n",
    "gencode_v47_path = \"/oak/stanford/groups/smontgom/dnachun/data/dgtex/dGTEx_Consortium_resources/NHP-dGTEx_references/Enhanced_GTFs/Macaques/v1/Macaque_Mmul10_enhanced_v1_genes_collapsed_stranded.gtf\"\n",
    "gencode_v47=pr.read_gtf(gencode_v47_path)\n",
    "gencode_v47_df = gencode_v47.as_df()\n",
    "gencode_v47_df = gencode_v47_df[~(gencode_v47_df['Chromosome'] == 'chrM')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_closest_background(exon, background_starts, background_ends):\n",
    "#     exon_length = exon['end'] - exon['start']\n",
    "#     idx_right = bisect.bisect_left(background_starts, exon['start'])\n",
    "#     # catch index too far left\n",
    "#     while idx_right < len(background_starts) and background_starts[idx_right] < exon['start']:\n",
    "#         idx_right += 1\n",
    "\n",
    "#     idx_left = idx_right - 1\n",
    "#     max_iter = len(background_starts)  # Safety limit\n",
    "#     iter_count = 0\n",
    "\n",
    "#     while iter_count < max_iter:\n",
    "#         iter_count += 1\n",
    "#         # Try right side\n",
    "#         if idx_right < len(background_starts):\n",
    "#             right_start, right_end = background_starts[idx_right], background_ends[idx_right]\n",
    "#             if right_end - right_start >= exon_length:\n",
    "#                 return([right_start, right_start + exon_length])\n",
    "        \n",
    "#         # Try left side\n",
    "#         if idx_left >= 0:\n",
    "#             left_start, left_end = background_starts[idx_left], background_ends[idx_left]\n",
    "#             if left_end - left_start >= exon_length:\n",
    "#                 return([left_end-exon_length, left_end])  \n",
    "        \n",
    "#         idx_right += 1\n",
    "#         idx_left -= 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# null_exons = []\n",
    "# for chrom, background_chrom in background.groupby('chr'):\n",
    "#     null_exons_chr = []\n",
    "#     print(chrom)\n",
    "#     background_starts = background_chrom['start'].values\n",
    "#     background_ends = background_chrom['end'].values\n",
    "#     exon_chrom = exons[exons['chr'] == chrom]\n",
    "#     print(len(exon_chrom))\n",
    "#     for idx, exon in tqdm(exon_chrom.iterrows(), total=len(exon_chrom)):\n",
    "#         null_exons_chr.append(get_closest_background(exon, background_starts, background_ends))\n",
    "#     null_exons_chr = pd.DataFrame(null_exons_chr, columns=['start', 'end'])\n",
    "#     null_exons_chr['exon_id'] = exon_chrom['exon_id'].values\n",
    "#     null_exons_chr['gene_id'] = exon_chrom['gene_id'].values\n",
    "#     null_exons_chr['chr'] = chrom\n",
    "#     null_exons.append(null_exons_chr)\n",
    "\n",
    "# # concat and save\n",
    "# null_exons_all_chrs = pd.concat(null_exons)\n",
    "# null_exons_all_chrs[['chr', 'start', 'end', 'exon_id', 'gene_id']].to_csv('/home/klawren/oak/gtex/data/other_references/nongenic_null/matched_nongenic_null_intergenic.bed', index=False, header=None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\"each gene have a single transcript whose id matches the parent gene id\" from rnaseqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # fix the null name to be compatible\n",
    "# null_exons_all_chrs['exon_id'] = null_exons_all_chrs['exon_id'].str.split(';').str[0]\n",
    "\n",
    "# # merge in the null information\n",
    "# null_exons_gtf = pd.merge(gencode_v47_df[gencode_v47_df['Feature'] == 'exon'].reset_index(), null_exons_all_chrs, on='exon_id', suffixes=['', '_null'])\n",
    "# # combine into the main gtf based on index \n",
    "# null_full_gtf = pd.merge(gencode_v47_df.reset_index(), null_exons_gtf[['index', 'start', 'end']], on='index', how='left', suffixes=['', '_null'])\n",
    "\n",
    "# # add gene starts and ends as bounds \n",
    "# gene_starts = null_full_gtf.groupby('gene_id').agg({'start':'min', 'end':'max'})\n",
    "# null_full_gtf = pd.merge(null_full_gtf, gene_starts, on='gene_id', how='left', suffixes=['', '_gene'])\n",
    "\n",
    "# # replace the start and end with the null version\n",
    "# null_full_gtf['Start'] = np.where(null_full_gtf['Feature'] == 'exon', null_full_gtf['start'], null_full_gtf['start_gene'])\n",
    "# null_full_gtf['End'] = np.where(null_full_gtf['Feature'] == 'exon', null_full_gtf['end'], null_full_gtf['end_gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Function to convert to gtf\n",
    "# def df_to_gtf(df, output_file):\n",
    "#     with open(output_file, 'w') as f:\n",
    "#         f.write(\"##description: nongenic nulls matched to exons from annotation of the human genome (GRCh38), version 47 (Ensembl 113)\\n##provider: Kate Lawrence\\n##contact: klawren@stanford.edu\\n##format: gtf\\n##date: 2025-03-21\")\n",
    "#         for index, row in df.iterrows():\n",
    "#             attributes = f'gene_id \"{row[\"gene_id\"]}\"; transcript_id \"{row[\"transcript_id\"]}\"; exon_id \"{row[\"exon_id\"]}\";'\n",
    "#             gtf_line = f'{row[\"Chromosome\"]}\\t{row[\"Source\"]}\\t{row[\"Feature\"]}\\t{row[\"Start\"]}\\t{row[\"End\"]}\\t{row[\"Score\"]}\\t{row[\"Strand\"]}\\t{row[\"Frame\"]}\\t{attributes}\\n'\n",
    "#             f.write(gtf_line)\n",
    "\n",
    "# df_to_gtf(null_full_gtf, '/home/klawren/oak/gtex/data/other_references/nongenic_null/matched_nongenic_null_intergenic.v47.gtf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### dont allow exon overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_closest_background_no_overlap(exon, background_starts, background_ends):\n",
    "    exon_length = exon['end'] - exon['start']\n",
    "    idx_right = bisect.bisect_left(background_starts, exon['start'])\n",
    "    # catch index too far left\n",
    "    while idx_right < len(background_starts) and background_starts[idx_right] < exon['start']:\n",
    "        idx_right += 1\n",
    "\n",
    "    idx_left = idx_right - 1\n",
    "    max_iter = len(background_starts)  # Safety limit\n",
    "    iter_count = 0\n",
    "\n",
    "    while iter_count < max_iter:\n",
    "        iter_count += 1\n",
    "        # Try right side\n",
    "        if idx_right < len(background_starts):\n",
    "            right_start, right_end = background_starts[idx_right], background_ends[idx_right]\n",
    "            if right_end - right_start >= exon_length:\n",
    "                background_starts[idx_right] = background_starts[idx_right] + exon_length\n",
    "                return([right_start, right_start + exon_length])\n",
    "        \n",
    "        # Try left side\n",
    "        if idx_left >= 0:\n",
    "            left_start, left_end = background_starts[idx_left], background_ends[idx_left]\n",
    "            if left_end - left_start >= exon_length:\n",
    "                background_ends[idx_left] = background_ends[idx_left] - exon_length\n",
    "                return([left_end-exon_length, left_end])  \n",
    "        \n",
    "        idx_right += 1\n",
    "        idx_left -= 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "null_exons = []\n",
    "for chrom, background_chrom in background.groupby('chr'):\n",
    "    null_exons_chr = []\n",
    "    print(chrom)\n",
    "    background_starts = background_chrom['start'].values\n",
    "    background_ends = background_chrom['end'].values\n",
    "    exon_chrom = exons[exons['chr'] == chrom]\n",
    "    print(len(exon_chrom))\n",
    "    for idx, exon in tqdm(exon_chrom.iterrows(), total=len(exon_chrom)):\n",
    "        null_exons_chr.append(get_closest_background_no_overlap(exon, background_starts, background_ends))\n",
    "    null_exons_chr = pd.DataFrame(null_exons_chr, columns=['start', 'end'])\n",
    "    null_exons_chr['exon_id'] = exon_chrom['exon_id'].values\n",
    "    null_exons_chr['gene_id'] = exon_chrom['gene_id'].values\n",
    "    null_exons_chr['chr'] = chrom\n",
    "    null_exons.append(null_exons_chr)\n",
    "\n",
    "# concat and save\n",
    "null_exons_all_chrs = pd.concat(null_exons)\n",
    "null_exons_all_chrs[['chr', 'start', 'end', 'exon_id', 'gene_id']].to_csv('/oak/stanford/groups/smontgom/klawren/nph_gtex/data/tmp/macaques/matched_nongenic_null_intergenic_no_overlap.bed', index=False, header=None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the null name to be compatible\n",
    "null_exons_all_chrs['exon_id'] = null_exons_all_chrs['exon_id'].str.split(';').str[0]\n",
    "\n",
    "# merge in the null information\n",
    "null_exons_gtf = pd.merge(gencode_v47_df[gencode_v47_df['Feature'] == 'exon'].reset_index(), null_exons_all_chrs, on='exon_id', suffixes=['', '_null'])\n",
    "# combine into the main gtf based on index \n",
    "null_full_gtf = pd.merge(gencode_v47_df.reset_index(), null_exons_gtf[['index', 'start', 'end']], on='index', how='left', suffixes=['', '_null'])\n",
    "\n",
    "# add gene starts and ends as bounds \n",
    "gene_starts = null_full_gtf.groupby('gene_id').agg({'start':'min', 'end':'max'})\n",
    "null_full_gtf = pd.merge(null_full_gtf, gene_starts, on='gene_id', how='left', suffixes=['', '_gene'])\n",
    "\n",
    "# replace the start and end with the null version\n",
    "null_full_gtf['Start'] = np.where(null_full_gtf['Feature'] == 'exon', null_full_gtf['start'], null_full_gtf['start_gene'])\n",
    "null_full_gtf['End'] = np.where(null_full_gtf['Feature'] == 'exon', null_full_gtf['end'], null_full_gtf['end_gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to convert to gtf\n",
    "def df_to_gtf(df, output_file):\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(\"##description: nongenic nulls matched to exons from annotation of the human genome (GRCh38), version 47 (Ensembl 113)\\n##provider: Kate Lawrence\\n##contact: klawren@stanford.edu\\n##format: gtf\\n##date: 2025-03-21\")\n",
    "        for index, row in df.iterrows():\n",
    "            attributes = f'gene_id \"{row[\"gene_id\"]}\"; transcript_id \"{row[\"transcript_id\"]}\"; exon_id \"{row[\"exon_id\"]}\";'\n",
    "            gtf_line = f'{row[\"Chromosome\"]}\\t{row[\"Source\"]}\\t{row[\"Feature\"]}\\t{row[\"Start\"]}\\t{row[\"End\"]}\\t{row[\"Score\"]}\\t{row[\"Strand\"]}\\t{row[\"Frame\"]}\\t{attributes}\\n'\n",
    "            f.write(gtf_line)\n",
    "\n",
    "df_to_gtf(null_full_gtf, '/oak/stanford/groups/smontgom/klawren/nph_gtex/references/macaques_null_intergenic.gtf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### matched null for v10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exons = pd.read_csv(\"/home/klawren/oak/gtex/data/other_references/nongenic_null_v10/exons_to_match.bed\", sep='\\t', header=None, names=['chr', 'start', 'end', 'exon_id', 'gene_id'])\n",
    "# exons['length'] = exons['end'] - exons['start']\n",
    "# exons = exons[~exons['chr'].isin(['chrM'])]\n",
    "\n",
    "# # if introns are allowed\n",
    "# #background = pd.read_csv(\"/home/klawren/oak/gtex/data/other_references/nongenic_null/background_regions.bed\", sep='\\t', header=None, names=['chr', 'start', 'end'])\n",
    "# # if introns aren't allowed \n",
    "# background = pd.read_csv(\"/home/klawren/oak/gtex/data/other_references/nongenic_null_v10/intergenic_regions.bed\", sep='\\t', header=None, names=['chr', 'start', 'end'])\n",
    "\n",
    "# background = background[background['chr'].isin(exons['chr'].unique())]\n",
    "# background = background.sort_values(['chr', 'start'])\n",
    "\n",
    "# # load in gencode 39 gtf\n",
    "# gencode_v39_path = \"/home/klawren/oak/gtex/data/realign_references/gencode.v39.genes.gtf\"\n",
    "# gencode_v39=pr.read_gtf(gencode_v39_path)\n",
    "# gencode_v39_df = gencode_v39.as_df()\n",
    "# gencode_v39_df = gencode_v39_df[~(gencode_v39_df['Chromosome'] == 'chrM')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# null_exons = []\n",
    "# for chrom, background_chrom in background.groupby('chr'):\n",
    "#     null_exons_chr = []\n",
    "#     print(chrom)\n",
    "#     background_starts = background_chrom['start'].values\n",
    "#     background_ends = background_chrom['end'].values\n",
    "#     exon_chrom = exons[exons['chr'] == chrom]\n",
    "#     print(len(exon_chrom))\n",
    "#     for idx, exon in tqdm(exon_chrom.iterrows(), total=len(exon_chrom)):\n",
    "#         null_exons_chr.append(get_closest_background_no_overlap(exon, background_starts, background_ends))\n",
    "#     null_exons_chr = pd.DataFrame(null_exons_chr, columns=['start', 'end'])\n",
    "#     null_exons_chr['exon_id'] = exon_chrom['exon_id'].values\n",
    "#     null_exons_chr['gene_id'] = exon_chrom['gene_id'].values\n",
    "#     null_exons_chr['chr'] = chrom\n",
    "#     null_exons.append(null_exons_chr)\n",
    "\n",
    "# # concat and save\n",
    "# null_exons_all_chrs = pd.concat(null_exons)\n",
    "# null_exons_all_chrs[['chr', 'start', 'end', 'exon_id', 'gene_id']].to_csv('/home/klawren/oak/gtex/data/other_references/nongenic_null_v10/matched_nongenic_null_intergenic_no_overlap.bed', index=False, header=None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # fix the null name to be compatible\n",
    "# null_exons_all_chrs['exon_id'] = null_exons_all_chrs['exon_id'].str.split(';').str[0]\n",
    "\n",
    "# # merge in the null information\n",
    "# null_exons_gtf = pd.merge(gencode_v39_df[gencode_v39_df['Feature'] == 'exon'].reset_index(), null_exons_all_chrs, on='exon_id', suffixes=['', '_null'])\n",
    "# # combine into the main gtf based on index \n",
    "# null_full_gtf = pd.merge(gencode_v39_df.reset_index(), null_exons_gtf[['index', 'start', 'end']], on='index', how='left', suffixes=['', '_null'])\n",
    "\n",
    "# # add gene starts and ends as bounds \n",
    "# gene_starts = null_full_gtf.groupby('gene_id').agg({'start':'min', 'end':'max'})\n",
    "# null_full_gtf = pd.merge(null_full_gtf, gene_starts, on='gene_id', how='left', suffixes=['', '_gene'])\n",
    "\n",
    "# # replace the start and end with the null version\n",
    "# null_full_gtf['Start'] = np.where(null_full_gtf['Feature'] == 'exon', null_full_gtf['start'], null_full_gtf['start_gene'])\n",
    "# null_full_gtf['End'] = np.where(null_full_gtf['Feature'] == 'exon', null_full_gtf['end'], null_full_gtf['end_gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Function to convert to gtf\n",
    "# def df_to_gtf(df, output_file):\n",
    "#     with open(output_file, 'w') as f:\n",
    "#         f.write(\"##description: nongenic nulls matched to exons from annotation of the human genome (GRCh38), version 39\\n##provider: Kate Lawrence\\n##contact: klawren@stanford.edu\\n##format: gtf\\n##date: 2025-03-28\")\n",
    "#         for index, row in df.iterrows():\n",
    "#             attributes = f'gene_id \"{row[\"gene_id\"]}\"; transcript_id \"{row[\"transcript_id\"]}\"; exon_id \"{row[\"exon_id\"]}\";'\n",
    "#             gtf_line = f'{row[\"Chromosome\"]}\\t{row[\"Source\"]}\\t{row[\"Feature\"]}\\t{row[\"Start\"]}\\t{row[\"End\"]}\\t{row[\"Score\"]}\\t{row[\"Strand\"]}\\t{row[\"Frame\"]}\\t{attributes}\\n'\n",
    "#             f.write(gtf_line)\n",
    "\n",
    "# df_to_gtf(null_full_gtf, '/home/klawren/oak/gtex/data/realign_references/gencode.v39.matched_nongenic_null_intergenic.gtf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
