{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### lncRNAs quantified? realive to background\n",
    "1. construct backgorund gtf\n",
    "2. map reads with rnaseqc\n",
    "3. compare to see if lncRNAs are quantified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import pyranges as pr\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### create background gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_v47_path = \"/home/klawren/oak/gtex/data/references/gencode.v47.annotation.gtf\"\n",
    "gencode_v47=pr.read_gtf(gencode_v47_path).as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11120</td>\n",
       "      <td>24894</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11120</td>\n",
       "      <td>14413</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11120</td>\n",
       "      <td>11211</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004248723.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004248735.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4105480</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004015123.1</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4105481</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004015124.1</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4105482</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00004015125.1</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4105483</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00004015126.1</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4105484</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00004015127.1</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4105485 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "0             chr1  HAVANA        gene     11120     24894     .      +     .   \n",
       "1             chr1  HAVANA  transcript     11120     14413     .      +     .   \n",
       "2             chr1  HAVANA        exon     11120     11211     .      +     .   \n",
       "3             chr1  HAVANA        exon     12009     12227     .      +     .   \n",
       "4             chr1  HAVANA        exon     12612     12721     .      +     .   \n",
       "...            ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "4105480       chrY  HAVANA        exon  57214349  57214397     .      -     .   \n",
       "4105481       chrY  HAVANA        exon  57213879  57213964     .      -     .   \n",
       "4105482       chrY  HAVANA        exon  57213525  57213602     .      -     .   \n",
       "4105483       chrY  HAVANA        exon  57213203  57213357     .      -     .   \n",
       "4105484       chrY  HAVANA        exon  57212183  57213125     .      -     .   \n",
       "\n",
       "                   gene_id               gene_type  ... exon_number  \\\n",
       "0        ENSG00000290825.2                  lncRNA  ...         NaN   \n",
       "1        ENSG00000290825.2                  lncRNA  ...         NaN   \n",
       "2        ENSG00000290825.2                  lncRNA  ...           1   \n",
       "3        ENSG00000290825.2                  lncRNA  ...           2   \n",
       "4        ENSG00000290825.2                  lncRNA  ...           3   \n",
       "...                    ...                     ...  ...         ...   \n",
       "4105480  ENSG00000292371.1  unprocessed_pseudogene  ...           1   \n",
       "4105481  ENSG00000292371.1  unprocessed_pseudogene  ...           2   \n",
       "4105482  ENSG00000292371.1  unprocessed_pseudogene  ...           3   \n",
       "4105483  ENSG00000292371.1  unprocessed_pseudogene  ...           4   \n",
       "4105484  ENSG00000292371.1  unprocessed_pseudogene  ...           5   \n",
       "\n",
       "                   exon_id transcript_support_level     havana_transcript  \\\n",
       "0                      NaN                      NaN                   NaN   \n",
       "1                      NaN                      NaN                   NaN   \n",
       "2        ENSE00004248723.1                      NaN                   NaN   \n",
       "3        ENSE00004248735.1                      NaN                   NaN   \n",
       "4        ENSE00003582793.1                      NaN                   NaN   \n",
       "...                    ...                      ...                   ...   \n",
       "4105480  ENSE00004015123.1                       NA  OTTHUMT00000058841.1   \n",
       "4105481  ENSE00004015124.1                       NA  OTTHUMT00000058841.1   \n",
       "4105482  ENSE00004015125.1                       NA  OTTHUMT00000058841.1   \n",
       "4105483  ENSE00004015126.1                       NA  OTTHUMT00000058841.1   \n",
       "4105484  ENSE00004015127.1                       NA  OTTHUMT00000058841.1   \n",
       "\n",
       "            hgnc_id           havana_gene          ont protein_id ccdsid  \\\n",
       "0               NaN                   NaN          NaN        NaN    NaN   \n",
       "1               NaN                   NaN          NaN        NaN    NaN   \n",
       "2               NaN                   NaN          NaN        NaN    NaN   \n",
       "3               NaN                   NaN          NaN        NaN    NaN   \n",
       "4               NaN                   NaN          NaN        NaN    NaN   \n",
       "...             ...                   ...          ...        ...    ...   \n",
       "4105480  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005        NaN    NaN   \n",
       "4105481  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005        NaN    NaN   \n",
       "4105482  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005        NaN    NaN   \n",
       "4105483  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005        NaN    NaN   \n",
       "4105484  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005        NaN    NaN   \n",
       "\n",
       "        artif_dupl  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  \n",
       "...            ...  \n",
       "4105480        NaN  \n",
       "4105481        NaN  \n",
       "4105482        NaN  \n",
       "4105483        NaN  \n",
       "4105484        NaN  \n",
       "\n",
       "[4105485 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_v47"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\"In order to use XBSeq for testing DE, we need to run HTSeq twice to measure the reads mapped to exonic regions (observed signal) and non-exonic regions (background noise). Firstly, we need to construct the gtf annotation file to measure the background noise:\n",
    "\n",
    "     \\item {Download refFlat table from UCSC database (http://genome.ucsc.edu) and create the preliminary list of gene-free regions,}\n",
    "\t \\item {Download tables of (a) all_mrna; (b) ensGene; (c) pseudoYale60Gene; (d) vegaGene;, (e)xenoMrna, and (f) xenoRefGene from UCSC database and remove \n",
    "\t regions appear in any of them from the gene-free regions,}\n",
    "     \\item {To guarantee gene-free regions are far enough from exonic regions, trim 100 bps from both sides of intronic regions and 1,000 bps from both sides \n",
    "\t of inter-genic regions,}\n",
    "     \\item {Shift each exon of a gene to the right nearest gene-free region.  Most of the shifted genes remain the same as the original structures of the genes,}\n",
    "     \\item {If the nearby gene-free region is too short, we may only preserve the exon size features but not the whole gene structure. The priority of shifting \n",
    "\t a region is: i) nearest right gene-free region, 2) nearest left gene-free region; 3) the second right nearest gene-free region and so on until the shift \n",
    "\t region of the original exon fits, and}\n",
    "     \\item {Shift each exon of a gene to the right nearest gene-free region.  Most of the shifted genes remain the same as the original structures of the genes,}\n",
    "     \\item {At last, we considered the shifted regions as the non-exonic regions for each gene and a final .gtf file was created}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
