# sample args
realign_bam_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/output/caudate_analysis/genome_bam
gtex_ids=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/data/other_references/caudate/caudate_samples.txt
reference_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/data/realign_references
reference_fasta=Homo_sapiens_assembly38_noALT_noHLA_noDecoy.fasta
chr_sizes=GRCh38.chrsizes
genes_gtf=gencode.v47.matched_nongenic_null_introns.gtf
intervals_bed=gencode.v47.GRCh38.insert_size_intervals_geq1000bp.bed
output_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/output/caudate_nongenic_null/v11_introns
code_dir=/oak/stanford/groups/smontgom/dnachun/data/gtex/v10/scripts/processing

